src/fluree/db/query/exec/where.cljc

```diff
@@ -463,49 +472,66 @@
       [nil o-fn*])
     [o o-fn]))
 
+(defn comparable-iri?
+  "When matching against"
+  [x]
+  (or (iri/sid? x) (nil? x)))
+
+(defn unmatched-optional-vars?
+  [triple-pattern]
+  (not-empty (keep get-optional triple-pattern)))
+
 (defn resolve-flake-range
-  [{:keys [t] :as db} tracker error-ch [s-mch p-mch o-mch]]
-  (let [s    (get-sid s-mch db)
-        s-fn (::fn s-mch)
-        p    (get-sid p-mch db)
-        p-fn (::fn p-mch)
-        o    (or (get-value o-mch)
-                 (get-sid o-mch db))
-        o-fn (::fn o-mch)
-        o-dt (some->> o-mch get-datatype-iri (iri/encode-iri db))
-
-        idx         (try* (index/for-components s p o o-dt)
-                          (catch* e
-                            (log/error e "Error resolving flake range")
-                            (async/put! error-ch e)))
-        [o* o-fn*]  (augment-object-fn db idx s p o o-fn)
-        start-flake (flake/create s p o* o-dt nil nil util/min-integer)
-        end-flake   (flake/create s p o* o-dt nil nil util/max-integer)
-        track-fuel  (track/track-fuel! tracker error-ch)
-        subj-filter (when s-fn
-                      (filter (fn [f]
-                                (-> unmatched
-                                    (match-subject db f)
-                                    s-fn))))
-        pred-filter (when p-fn
-                      (filter (fn [f]
-                                (-> unmatched
-                                    (match-predicate db f)
-                                    p-fn))))
-        obj-filter  (when o-fn*
-                      (filter (fn [f]
-                                (-> unmatched
-                                    (match-object db f)
-                                    o-fn*))))
-        flake-xf    (->> [subj-filter pred-filter obj-filter track-fuel]
-                         (remove nil?)
-                         (apply comp))
-        opts        {:idx         idx
-                     :to-t        t
-                     :start-flake start-flake
-                     :end-flake   end-flake
-                     :flake-xf    flake-xf}]
-    (query-range/resolve-flake-slices db tracker idx error-ch opts)))
+  [{:keys [t] :as db} tracker error-ch [s-mch p-mch o-mch :as triple-pattern]]
+  (let [s    (or (get-sid s-mch db)
+                 (get-value s-mch))
```

Comment:
anything in the `s` or `p` position should only be sids, so we should never consider `get-value` for `s` or `p`. i think we should instead short circuit if either `s` or `p` returns `nil` for `get-sid` and non `nil` for `get-value`.

Reply:
That's basically what I'm checking here, I just use `get-value` here so I can deal with all the short-circuiting logic in one place, on line 490.

Reply:
I think I would prefer line 488 be just `s (get-sid s-mch db)` and lin 495 be `(and (nil? s) (some? get-value s-mch))` to make it clear that there should only be an sid in the `s` position or we short circuit.

I think this is the less confusing thing to my eyes when scanning the file, but if you disagree strongly, then feel free to leave this as is.

Reply:
My thought is that it is perfectly valid to have a `get-value` match component for `s` or `p` - that is explicitly ok with the query language semantics and I don't want to give any impression that a non-iri value there is anomalous in any way. A non-iri value will generate no new matches, though, so that's why we check for comparability below.